### PR TITLE
[luci-interpreter]Add Uint8 Fullyconnected kernel.

### PR DESCRIPTION
**Pull request by struss**
This commit add Uint8 case `Fullyconnected` kernel.

ONE-DCO-1.0-Signed-off-by: KiDeuk Bang <rrstrous@nate.com>

### CODE REVIEW

File: compiler/luci-interpreter/src/kernels/FullyConnected.test.cpp [hunk 1]
```diff
@@ -26,26 +26,56 @@ namespace
 
 using namespace testing;
 
-TEST(FullyConnectedTest, Float)
+// Returns the corresponding Bias DataType given the type T.
+template <typename T> constexpr DataType getBiasType()
```
Comment (s-barannikov) [inline compiler/luci-interpreter/src/kernels/FullyConnected.test.cpp:30]: There is `getElementType` in `TestUtils.h`, please use it.

File: compiler/luci-interpreter/src/kernels/FullyConnected.cpp [hunk 2]
```diff
@@ -36,27 +36,53 @@ FullyConnected::FullyConnected(const Tensor *input, const Tensor *weights, const
...
-  assert(bias() == nullptr || bias()->element_type() == DataType::FLOAT32);
+  if (weights()->element_type() == DataType::U8 || weights()->element_type() == DataType::S8)
+  {
+    LUCI_INTERPRETER_CHECK(input()->element_type() == DataType::U8 ||
+                           input()->element_type() == DataType::S8);
+    LUCI_INTERPRETER_CHECK(output()->element_type() == DataType::U8 ||
+                           output()->element_type() == DataType::S8);
+    LUCI_INTERPRETER_CHECK(!bias() || bias()->element_type() == DataType::S32 ||
+                           bias()->element_type() == DataType::S64)
+  }
```
Comment (s-barannikov) [inline compiler/luci-interpreter/src/kernels/FullyConnected.cpp:47]: These checks do not correspond to the implementation. `evalQuantized` only works with U8 input / output / weights and S32 bias.

File: compiler/luci-interpreter/src/kernels/FullyConnected.cpp [hunk 3]
```diff
@@ -75,5 +101,40 @@ void FullyConnected::evalFloat() const
...
+
+  tflite::FullyConnectedParams op_params{};
+  op_params.input_offset = input_offset;
+  op_params.weights_offset = filter_offset;
+  op_params.output_offset = output_offset;
+  op_params.output_multiplier = output_multiplier;
+  op_params.output_shift = output_shift;
+  op_params.quantized_activation_min = output_activation_min;
+  op_params.quantized_activation_max = output_activation_max;
+  // op_params.lhs_cacheable = IsConstantTensor(filter);
```
Comment (s-barannikov) [inline compiler/luci-interpreter/src/kernels/FullyConnected.cpp:129]: Please do not leave commented out code.

File: compiler/luci-interpreter/src/kernels/FullyConnected.test.cpp (hunk 1 above)
Comment (struss) [reply compiler/luci-interpreter/src/kernels/FullyConnected.test.cpp:30]: On this, only float input case return for float, and quantize case for S32. so, I think this is little bit different.

File: compiler/luci-interpreter/src/kernels/FullyConnected.cpp (hunk 3 above)
Comment (struss) [reply compiler/luci-interpreter/src/kernels/FullyConnected.cpp:129]: Okay, I will fix it!

File: compiler/luci-interpreter/src/kernels/FullyConnected.test.cpp (hunk 1 above)
Comment (s-barannikov) [reply compiler/luci-interpreter/src/kernels/FullyConnected.test.cpp:30]: I see. In this case the comment shold tell about that (that the given type T is the input type, not the bias type).

File: compiler/luci-interpreter/src/kernels/FullyConnected.cpp (hunk 2 above)
Comment (struss) [reply compiler/luci-interpreter/src/kernels/FullyConnected.cpp:47]: Okay, I will fix it!

File: compiler/luci-interpreter/src/kernels/FullyConnected.test.cpp (hunk 1 above)
Comment (struss) [reply compiler/luci-interpreter/src/kernels/FullyConnected.test.cpp:30]: Sorry for confusing. I will update more clearly.

File: compiler/luci-interpreter/src/kernels/FullyConnected.test.cpp [hunk 4]
```diff
@@ -26,26 +26,54 @@ namespace
...
+           std::initializer_list<float> input_data, std::initializer_list<float> weights_data,
+           std::initializer_list<float> bias_data, std::initializer_list<float> output_data)
+{
+  float kQuantizedTolerance = getTolerance(-127, 128, 255);
+  std::pair<float, int32_t> quant_param = quantizationParams<T>(-127, 128);
+  Tensor input_tensor{
+      getElementType<T>(), input_shape, {{quant_param.first}, {quant_param.second}}, ""};
+  Tensor weights_tensor{
+      getElementType<T>(), weights_shape, {{quant_param.first}, {quant_param.second}}, ""};
+  Tensor bias_tensor{getBiasType<T>(), bias_shape, {}, ""};
```
Comment (jinevening) [inline compiler/luci-interpreter/src/kernels/FullyConnected.test.cpp:51]: How did you set the quantization params and tolerance?

Comment (struss) [reply compiler/luci-interpreter/src/kernels/FullyConnected.test.cpp:51]: Referenced from TFlite test, but just I reference that could cover all input and output data.
https://github.com/tensorflow/tensorflow/blob/ee598066c4cb31ec5ed3106e61ba99ef004a4bae/tensorflow/lite/kernels/fully_connected_test.cc#L460-L464

File: compiler/luci-interpreter/src/kernels/Utils.h [hunk 5]
```diff
@@ -94,6 +94,14 @@ void quantizeMultiplierSmallerThanOneExp(double double_multiplier, int32_t *quan
 
 Shape calculateShapeForBroadcast(const Shape &input1_shape, const Shape &input2_shape);
 
+inline void getQuantizedConvolutionMultipler(float input_scale, float filter_scale,
```
Comment (s-barannikov) [inline compiler/luci-interpreter/src/kernels/Utils.h:97]: Why don't return multiplier by value?

Comment (struss) [reply compiler/luci-interpreter/src/kernels/Utils.h:97]: When I refer from this function in TFlite, this function need to return 2 value(State code and multiplier). and while make a change for our interpreter, I just removed state return code and make return void. I think this is my mistake. I will fix it.

File: compiler/luci-interpreter/src/kernels/FullyConnected.test.cpp (hunk 4 above)
Comment (jinevening) [reply compiler/luci-interpreter/src/kernels/FullyConnected.test.cpp:51]: It seems that qparam of input and weights should be changed; please create another qparam with min/max -63.5/64.

Bias has qparam whose scale is (scale of input) * (scale of weights) and zero point is 0.
(https://github.com/tensorflow/tensorflow/blob/ee598066c4cb31ec5ed3106e61ba99ef004a4bae/tensorflow/lite/kernels/fully_connected_test.cc#L168)

Output uses the current qparam (-127~128), so leave it as is.

Comment (struss) [reply compiler/luci-interpreter/src/kernels/FullyConnected.test.cpp:51]: Okay I will fix it.

**Review by jinevening — APPROVED** [review]
LGTM

**Review by s-barannikov — APPROVED** [review]
LGTM

**Comment by jinevening** [comment]
@seanshpark Can you merge this PR?

**Review by seanshpark — APPROVED** [review]
LGTM

**Pull request merged**